### ISSUE TITLE
Add a flag which avoids using the validation data for creating a vocabulary.

### ISSUE THREAD

**Issue by kmkurn** [allenai/allennlp#516]
Hi, I was looking at the code for `train` command and noticed that the vocabulary is built from all datasets (train, dev, test). Is this intentional? I am asking because I thought the common thing to do is to build vocabulary from train set only so that words in dev/test set that does not occur in train set are converted into `<unk>` and given a single embedding. Building vocabulary from all three sets means that there will be no unknown words in dev/test set (assuming `min_count=1`).

**Comment by DeNeutoy** [comment]
Hi, yes this is intended. Basically it is an easy way to use pre-trained word vectors for unseen words in the test set - by constructing the vocabulary from the full dataset, we add the test-set words and find pretrained vectors for them, but they are never updated as they are never seen at training time. Then at test time, we can use those vectors without having to modify the embedding layer, which would be a real pain.

**Comment by kmkurn** [comment]
I see. So, if the word `foo` occurs once in the train set and twice in the dev/test set, and `min_count=2`, then `foo` will be included in the vocabulary because it occurs 3 times in total. Is there any way to not count the occurrence of `foo` in the dev/test set when building the vocab (aside from writing a custom `train` command)? I'd like for all words that occur fewer than `min_count` in the train set to be assigned a single `<unk>` embedding.

**Comment by DeNeutoy** [comment]
This is a little annoying, but you can do the following:

run `python -m allennlp/run.py train /your/config -s /some/output/path` where your config only contains a `train_dataset_path`. This will save a directory `vocabulary` in `/some/output/path`, which will contain a vocabulary built from only the training data.

The `from_params` [method](https://github.com/allenai/allennlp/blob/master/allennlp/data/vocabulary.py#L356) in the `Vocabulary` class can handle a directory path to a pre-existing vocabulary. Now you can add your evaluation data and testing data in, along with a `directory_path` key in the `vocabulary` section of your config, meaning you will be using a vocabulary built only from the training data. You can also manually create this vocabulary directory if you need to make other modifications/test other data splits.

**Comment by kmkurn** [comment]
Thank you, I'll try that.

**Comment by matt-gardner** [comment]
We should probably add a flag to not use the validation data for creating vocab.  I'm opening this issue again until we either add that flag or make a new issue specifically for adding the flag.

**Comment by janLuke** [comment]
My (unrequested) two cents. If you use the entire dataset to build the vocabulary, of course your test set is going to have a smaller percentage of OOV tokens than normal and that is a bias. It is probably very small, given that OOV tokens are going to be very rare tokens for any reasonably-sized vocabulary, but it's there.

Sure, you can always augment your ``Embedding`` layer to include other pre-trained word vectors from the same set you used during training. Okay, but you should probably be aware of the fact that your test performance is not biased only if you actually do that before you use your model with new data (if possible) or you use the entire set of pre-trained embeddings after training. It's a biased performance wrt the particular instance of ``Model`` you trained. 

Does it make sense?

(*) In practice, I guess the bias quickly decreases with the amount of pre-trained vectors (sorted by frequency) included in the embedding matrix.
